velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -184,6 +201,78 @@ void WindowFuzzer::go(const std::string& /*planPath*/) {
   VELOX_NYI();
 }
 
+void WindowFuzzer::testAlternativePlans(
+    const std::vector<std::string>& partitionKeys,
+    const std::vector<SortingKeyAndOrder>& sortingKeysAndOrders,
+    const std::string& frame,
+    const std::string& functionCall,
+    const std::vector<RowVectorPtr>& input,
+    bool customVerification,
+    const velox::test::ResultOrError& expected) {
+  std::vector<AggregationFuzzerBase::PlanWithSplits> plans;
+
+  std::vector<std::string> allKeys;
+  for (const auto& key : partitionKeys) {
+    allKeys.push_back(key + " NULLS FIRST");
```

Comment:
nit should we also do a NULLS LAST with 50% probability.

Reply:
@kgpai : NULLS LAST is the default so there isn't a need to add it specifically. Adding it only tests that NULLS LAST is parsed correctly which should be covered by unit tests.

Reply:
Thanks @aditi-pandit

Reply:
Would it make sense still to alternate some keys with NULLS FIRST and some NULLS LAST just to see if that catches any bugs ?

Reply:
Hi @kgpai, "NULLS FIRST" here is needed because allKeys is used in the streaming window plans where we want to order the inputs by the same order that a regular non-streaming window operation would perform. When a regular non-streaming window operation performs "partition-by", it places nulls first (found by experiment), so here we add "nulls first" to all partition keys.

Reply:
>  it places nulls first (found by experiment), so here we add "nulls first" to all partition keys. 

Thanks , can you add that as a comment .

Reply:
Hi [@kgpai](https://github.com/kgpai), "NULLS FIRST" here is needed because allKeys is used in the streaming window plans where we want to order the inputs by the same order that a regular non-streaming window operation would perform. When Velox regular non-streaming window operation performs "partition-by", it places nulls first, so here we add "nulls first" to all partition keys to make the steaming window plan produce the same result as non-streaming window plan.

Reply:
Good question. I used NULLS FIRST here because I noticed that when I used NULLS LAST, the streaming window plans produce different results from the non-streaming plans. But thinking about it twice, I feel NULLS LAST shouldn't affect the result. It may be a bug. I'm investigating.

Reply:
@kagamiori, @kgpai  : My bad... I misread this. 

So here the NULLS FIRST Is added to partition keys. In Sort Build window we order all the rows by (partition + order by) keys to separate partitions. To do this we use NULLS FIRST in the sorting. So the partition with NULL keys gets output in the beginning. 

The Window operator doesn't dictate if the NULL partitions should come before or not. Using NULLS FIRST Is just an implementation detail of the sorting. We had an option to chose either one. There was no specific meaning attached to it.

In StreamingWindowBuild we just go with the order of the partition rows sent to it. 

So for matching results you need to add NULLS FIRST for streaming partitions in the same order. NULLS LAST will not work.

Reply:
Thanks again @aditi-pandit .

Reply:
Thanks, @aditi-pandit, for helping explaining! So I investigated the implementation of StreamingWindowBuild and basically found the same thing. StreamingWindowBuild assumes that nulls in the partition-by columns are placed first in the ordered input (through WindowBuild::partitionKeyInfo_). If nulls are actually placed at the end in the input, StreamingWindowBuild creates incorrect partitions, hence causing incorrect window operation result.